src/server/DefaultRequestInformationExtractor.ts

```diff
@@ -0,0 +1,182 @@
+            // was application/graphql, parse the string body.
+            if (typeof body === 'string' && typeInfo.type === 'application/graphql') {
+                return { query: body };
+            }
+
+            // Already parsed body we didn't recognise? Parse nothing.
+            if (body != null) {
+                return {};
+            }
+
+            const rawBody = await this.readBody(request, typeInfo);
+
+            if (typeof rawBody ==='string') {
+                // Use the correct body parser based on Content-Type header.
+                switch (typeInfo.type) {
+                case 'application/graphql':
+                    return { query: rawBody };
+                case 'application/json':
+                    try {
+                        return JSON.parse(rawBody);
+                    } catch {
+                        return {
+                            error: { graphQLError: new GraphQLError('POST body contains invalid JSON.'), statusCode: 400 }
+                        }
+                    }
+                case 'application/x-www-form-urlencoded':
+                    return this.extractInformationFromUrlParameters(`host?${rawBody}.`)
+                }
+
+                // If no Content-Type header matches, parse nothing.
+                return {
+                    error: { graphQLError: new GraphQLError(`POST body contains invalid content type: ${typeInfo.type}.`), statusCode: 400 }
+                }
+            } else {
+                return {
+                    error: rawBody
+                }
+            }
+        } catch {
+            return {
+                error: { graphQLError: new GraphQLError('Content type could not be parsed.'), statusCode: 400 }
+            }
+        }
+    }
+
+    // Read and parse a request body.
+    async readBody(
+        req: Request,
+        typeInfo: ParsedMediaType,
+    ): Promise<string | GraphQLErrorWithStatusCode> {
+        const charset = typeInfo.parameters.charset?.toLowerCase() ?? 'utf-8';
+
+        // Assert charset encoding per JSON RFC 7159 sec 8.1
+        if (charset !== 'utf8' && charset !== 'utf-8' && charset !== 'utf16le') {
+            return {
+                graphQLError: new GraphQLError(`Unsupported charset "${charset.toUpperCase()}".`),
+                statusCode: 415
+            }
+        }
+
+        // Get content-encoding (e.g. gzip)
+        const contentEncoding = req.headers['content-encoding'];
+        const encoding =
+            typeof contentEncoding === 'string'
+                ? contentEncoding.toLowerCase()
+                : 'identity';
+        const maxBuffer = 100 * 1024; // 100kb
+        const stream = this.decompressed(req, encoding);
+        if ('graphQLError' in stream && 'statusCode' in stream) {
+            return {
+                graphQLError: stream.graphQLError,
+                statusCode: stream.statusCode
+            }
+        }
+
+        // Read body from stream.
```

Comment:
Maybe instead of commenting this block of code, refactor it into a small private function?

Reply:
Extracted to own function